compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs

```diff
@@ -265,9 +265,24 @@ namespace std
                 getAllocator _ = mempty
 
         -- move constructor
-        moveCtor = CPP.ifndef CPP.rvalueReferences [lt|
-        #{declName}(#{declName}&&#{param})#{initList}#{ctorBody}|]
+        moveCtor = if hasMetaFields then guardedDefine else implicitlyDeclared
           where
+            guardedDefine = CPP.ifndef CPP.rvalueReferences define
+
+            -- default OK when there are no meta fields, but fall back to
+            -- generated one for compilers that don't support = default for
+            -- move constructors
+            implicitlyDeclared = [lt|
+#if !defined(#{CPP.defaultedMoveCtors})
+        #{declName}(#{declName}&& other) = default;
+#elif !defined(#{CPP.rvalueReferences})
+        #{define}
+#endif|]
+
+            -- define ctor to perform member-by-member move and--if
+            -- needed--initialize meta fields
+            define = [lt|
+        #{declName}(#{declName}&&#{param})#{initList}#{ctorBody}|]
```

Comment:
I think the names don't help with understanding the logic, e.g. `implicitlyDeclared` in fact generates a conditional implicit/explicit ctor. Consider something like this:

```
moveCtor = if hasMetaFields then [lt|
#if !defined(#{CPP.rvalueReferences})
        #{explicit}
#endif
|]
                            else lt|
#if !defined(#{CPP.defaultedMoveCtors})
        #{implicit}
#elif !defined(#{CPP.rvalueReferences})
        #{explicit}
#endif
|]

```